[quant] Add dequantize.tensors

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35130 [quant][graphmode] Make interpolate/upsample work again
* #35142 [quant][graphmode] SwapDeQuant support prim::If
* #35141 [quant][graphmode][refactor] getGeneralOpTensorInputIndexes -> getGeneralOpTensorInputs
* #35135 [quant][graphmode][refactor] swapDeQuant takes block as arugment
* #35077 [quant][graphmode] Fold quantized prepacking ops
* #35073 [quant] Make conv2d_prepack and linear_prepack pure
* #34855 [quant][graphmode] quantization support for prim::CallFunction
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34854 [quant][graphmode] quantization support for view, transpose, contiguos
* #34806 [quant][graphmode] quantization support for aten::chunk
* #34807 [quant][graphmode] quantization support for prim::ListUnpack
* #34804 [quant][graphmode] Replicate quantize node for prim::If
* #34572 [quant][graphmode] quantization support for aten::add
* #34518 [quant][graphmode] quantization work for prim::If
* #34571 [quant][graphmode][fix] use observed_values_ to check values are observed
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* **#34348 [quant] Add dequantize.tensors**

Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

jerryzh168 committed:
[quant] Add dequantize.tensors

Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -32,6 +32,14 @@ Tensor dequantize_quant(const Tensor& self) {
   return get_qtensorimpl(self)->quantizer()->dequantize(self);
 }
 
+std::vector<Tensor> dequantize_tensors_quant(TensorList tensors) {
```
Why is this needed? Can we handle lists in a more primitive manner? What if we need to add a dict of tensors, would we add one more function for this purpose?

In general, how are lists of tensors handled? How do we observe a list of tensors and quantize them?

jerryzh168 replied:
explained offline

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3601,12 +3601,17 @@
   dispatch:
     QuantizedCPU: dequantize_quant
 
+- func: dequantize.tensors(Tensor[] tensors) -> Tensor[]
```
Why is ListConstruct special? What about other structures like Tuple or Dict of tensors?

jerryzh168 replied:
we'll need to handle those as well if they come up in real use cases, but by far I didn't see that. Reason we have lists are mostly because we have quantized ops like cat that takes list as input. if we have quantized ops that takes Dict or Tuple as inputs we'll probably need to add support for these other types as well

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -32,6 +32,14 @@ Tensor dequantize_quant(const Tensor& self) {
 
+std::vector<Tensor> dequantize_tensors_quant(TensorList tensors) {
+  std::vector<Tensor> dequantized_tensors;
+  for (auto i = 0; i < tensors.size(); ++i) {
```
can we use parallel for here?

jerryzh168 replied:
I would delay this optimization until there is a need, most likely this function will never be run since it will be fused into the quantized::cat pattern.

z-a-f reviewed (approved):
LGTM

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -32,6 +32,14 @@ Tensor dequantize_quant(const Tensor& self) {
   return get_qtensorimpl(self)->quantizer()->dequantize(self);
 }
 
+std::vector<Tensor> dequantize_tensors_quant(TensorList tensors) {
```
Don't return a std::vector, return a TensorList instead (TensorList == torch::List<Tensor>). std::vector requires a copy behind the scenes.

jerryzh168 replied:
I think that is not the type generated by native_functions.yaml, I looked up another function that returns Tensor[] and it says:
```
795:std::vector<Tensor> meshgrid(TensorList tensors); // {"schema": "aten::meshgrid(Tensor[] tensors) -> Tensor[]", "compound": "true"}
```

jerryzh168 commented on torch/_torch_docs.py:
```diff
@@ -1513,6 +1513,20 @@ def merge_dicts(*dicts):
+
+Given a list of quantized Tensors, dequantize them and return a list of dequantized Tensors
+""".format(**reduceops_common_args))
+
```
This seems impossible, what should I do here? @smessmer ?

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize.tensors"


Summary:
We need this function to do swap dequantize for prim::ListConstruct since
the output of prim::ListConstruct is a list of Tensors

Test Plan:
.
Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20504454](https://our.internmc.facebook.com/intern/diff/D20504454)

[ghstack-poisoned]
